File: src/dart/lib/news.dart
library dutapi;

import 'dart:developer';

import 'package:html/parser.dart';
import 'package:http/http.dart' as http;

import 'model/global_obj_var.dart';
import 'model/news_obj.dart';

class News {
  static Future<List<NewsGlobal>> _getNews(
      {int page = 1, required NewsType newsType}) async {
    List<NewsGlobal> result = [];

    String newsUrl =
        "http://sv.dut.udn.vn/WebAjax/evLopHP_Load.aspx?E=${(newsType == NewsType.global) ? 'CTRTBSV' : 'CTRTBGV'}&PAGETB=$page&COL=TieuDe&NAME=&TAB=0";

    try {
      final response = await http.Client().get(Uri.parse(newsUrl));
      if (response.statusCode == 200) {
        var doc = parse(response.body).getElementById('pnBody');
        if (doc != null) {
          doc.getElementsByClassName('tbBox').forEach((element) {
            NewsGlobal newsItem = NewsGlobal();
            var splitted = element
                .getElementsByClassName('tbBoxCaption')[0]
                .text
                .split(':     ');
            // Date
            final dateTime = DateTime.parse(
                splitted[0].split('/').reversed.toList().join('-'));
            newsItem.date = dateTime.millisecondsSinceEpoch;
            // Title
            newsItem.title = splitted[1];
            // Html and content string
            newsItem.content =
                element.getElementsByClassName('tbBoxContent')[0].innerHtml;
            newsItem.contentString =
                element.getElementsByClassName('tbBoxContent')[0].text;
            // Detect links
            int position = 1;
            var contentTemp = newsItem.contentString;
            element
                .getElementsByClassName('tbBoxContent')[0]
                .getElementsByTagName('a')
                .forEach((element) {
              if (contentTemp.contains(element.text)) {
                position += contentTemp.indexOf(element.text);
                NewsLinkItem newsLink = NewsLinkItem(
                    text: element.text,
                    position: position,
                    url: element.attributes['href']!);
                newsItem.links.add(newsLink);

                position += element.text.length;

                // https://stackoverflow.com/questions/24220509/exception-when-replacing-brackets
                // var temp = contentTemp.split(element.text);
                // if (temp.length > 1) contentTemp = temp[1];
                contentTemp = contentTemp.substring(
                    contentTemp.indexOf(element.text) + element.text.length);
              }
            });

            // Add to list.
            result.add(newsItem);
          });
        }
      }
    } catch (ex) {
      log(ex.toString());
      result.clear();
      result = List.empty();
    }
    return result;
  }

  static Future<List<NewsGlobal>> getNewsGlobal({int page = 1}) async {
    return await _getNews(page: page, newsType: NewsType.global);
  }

  static Future<List<NewsSubject>> getNewsSubject({int page = 1}) async {
    List<NewsSubject> result = [];

    try {
      await _getNews(page: page, newsType: NewsType.subject).then((value) {
        for (var newsItem in value) {
          NewsSubject item = NewsSubject();
          // Add all items in news global.
          item.date = newsItem.date;
          item.title = newsItem.title;
          item.content = newsItem.content;
          item.contentString = newsItem.contentString;
          item.links.clear();
          item.links.addAll(newsItem.links);

          // For title
          String lecturerProcessing =
              newsItem.title.split(' thông báo đến lớp: ')[0].trim();

          // Lecturer name
          item.lecturerName =
              lecturerProcessing.substring(lecturerProcessing.indexOf(' ') + 1);
          // Lecturer gender
          switch (lecturerProcessing
              .substring(0, lecturerProcessing.indexOf(' '))
              .trim()
              .toLowerCase()) {
            case 'thầy':
              {
                item.lecturerGender = LecturerGender.male;
                break;
              }
            case 'cô':
              {
                item.lecturerGender = LecturerGender.female;
                break;
              }
            default:
              {
                item.lecturerGender = LecturerGender.other;
                break;
              }
          }

          // Subject processing
          newsItem.title.split(' thông báo đến lớp: ')[1].split(' , ').forEach(
            (element) {
              final start = element.lastIndexOf('[') + 1;
              final end = element.lastIndexOf(']');
              final classId = element.substring(start, end);
              final className = element.substring(0, start - 1).trimRight();

              final affectedClassIndex = item.affectedClasses
                  .indexWhere((element) => element.subjectName == className);
              if (affectedClassIndex > -1) {
                var affectedClass = item.affectedClasses[affectedClassIndex];
                affectedClass.codeList.add(
                  SubjectCodeItem.fromTwoLastDigit(
                    studentYearId: classId.split('.')[0],
                    classId: classId.split('.')[1],
                  ),
                );
              } else {
                SubjectGroupItem subjectGroupItem = SubjectGroupItem();
                subjectGroupItem.subjectName = className;
                try {
                  subjectGroupItem.codeList.add(
                    SubjectCodeItem.fromTwoLastDigit(
                      studentYearId: classId.split('.')[0],
                      classId: classId.split('.')[1],
                    ),
                  );
                } catch (ex) {
                  subjectGroupItem.codeList.add(
                    SubjectCodeItem.fromTwoLastDigit(
                      studentYearId: classId.substring(0, 2),
                      classId: classId.substring(2),
                    ),
                  );
                }
                item.affectedClasses.add(subjectGroupItem);
              }
            },
          );

          // Check if is make up or leaving subject lessons.
          if (item.contentString.contains('HỌC BÙ')) {
            item.lessonStatus = LessonStatus.makeUp;
          } else if (item.contentString.contains('NGHỈ HỌC')) {
            item.lessonStatus = LessonStatus.leaving;
          } else {
            item.lessonStatus = LessonStatus.unknown;
          }

          // Date (These works apply only if lesson status is leaving and make up)
          if ([LessonStatus.leaving, LessonStatus.makeUp]
              .contains(item.lessonStatus)) {
            RegExp regExp = RegExp('\\d{2}[-|/]\\d{2}[-|/]\\d{4}');
            var firstMatch = regExp.firstMatch(item.contentString);
            if (firstMatch != null) {
              final dateTime = DateTime.parse(item.contentString
                  .substring(firstMatch.start, firstMatch.end)
                  .split('/')
                  .reversed
                  .toList()
                  .join('-'));
              item.affectedDate = dateTime.millisecondsSinceEpoch;
            }
          }

          // Lesson (These works apply only if lesson status is leaving and make up)
          if ([LessonStatus.leaving, LessonStatus.makeUp]
              .contains(item.lessonStatus)) {
            var query = '';
            if (item.lessonStatus == LessonStatus.makeUp) {
              query = 'tiết: .*[0-9],';
            } else {
              query = '\\(tiết:.*[0-9]\\)';
            }
            RegExp regExp = RegExp(query);
            var firstMatch =
                regExp.firstMatch(item.contentString.toLowerCase());
            if (firstMatch != null) {
              var splitted = item.contentString
                  .substring(firstMatch.start, firstMatch.end)
                  .toLowerCase()
                  .replaceFirst(
                      (item.lessonStatus == LessonStatus.makeUp)
                          ? 'tiết: '
                          : '(tiết:',
                      '')
                  .replaceFirst(
                      (item.lessonStatus == LessonStatus.makeUp) ? ',' : ')',
                      '')
                  .trim()
                  .split('-');
              item.affectedLessons.start = int.parse(splitted[0]);
              item.affectedLessons.end = int.parse(splitted[1]);
            }
          }

          // Room (These works apply only if lesson status is make up)
          if ([LessonStatus.makeUp].contains(item.lessonStatus)) {
            RegExp regExp = RegExp('phòng:.*');
            var firstMatch =
                regExp.firstMatch(item.contentString.toLowerCase());
            if (firstMatch != null) {
              item.affectedRoom = item.contentString
                  .toLowerCase()
                  .substring(firstMatch.start, firstMatch.end)
                  .replaceFirst('phòng:', '')
                  .replaceFirst(',', '')
                  .trim()
                  .toUpperCase();
            }
          }

          // Add to list
          result.add(item);
        }
      });
    } catch (ex) {
      log(ex.toString());
      log(ex.hashCode.toString());
      result.clear();
      result = List.empty();
    }
    return result;
  }
}

File: src/dotnet/DutWrapper.UWP/News.cs
using DutWrapper.UWP.Model;
using DutWrapper.UWP.Model.News;
using HtmlAgilityPack;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net;
using System.Net.Http;

namespace DutWrapper.UWP
{
    public static partial class News
    {
        private static List<NewsGlobalItem> GetNews(NewsType newsType, int page = 1, string query = null)
        {
            if (page < 1)
                throw new ArgumentException($"Page must be greater than 0! (current is {page})");

            List<NewsGlobalItem> result = new List<NewsGlobalItem>();

            try
            {
                HttpClient client = new HttpClient();
                client.BaseAddress = new Uri("http://sv.dut.udn.vn");

                HttpResponseMessage response = client.GetAsync($"/WebAjax/evLopHP_Load.aspx?E={(newsType == NewsType.Global ? "CTRTBSV" : "CTRTBGV")}&PAGETB={(page > 0 ? page : 1)}&COL=TieuDe&NAME={query}&TAB=1").Result;
                if (!response.IsSuccessStatusCode)
                    throw new Exception(String.Format("The request has return code {0}.", response.StatusCode));

                HtmlDocument htmlDoc = new HtmlDocument();
                htmlDoc.LoadHtml(response.Content.ReadAsStringAsync().Result);

                HtmlNodeCollection htmlDocNews = htmlDoc.DocumentNode.SelectNodes("//div[@class='tbBox']");

                // TODO: Add exception here.
                if (htmlDocNews == null || htmlDocNews.Count == 0)
                    throw new Exception($"No datas from sv.dut.udn.vn in page {page}.");

                foreach (HtmlNode htmlItem in htmlDocNews)
                {
                    NewsGlobalItem item = new NewsGlobalItem();

                    var htmlTemp = new HtmlDocument();
                    htmlTemp.LoadHtml(htmlItem.InnerHtml);

                    string title = htmlTemp.DocumentNode.SelectNodes("//div[@class='tbBoxCaption']")[0].InnerText;
                    string[] titleTemp = title.Split(new string[] { ":&nbsp;&nbsp;&nbsp;&nbsp; " }, StringSplitOptions.None);

                    if (titleTemp.Length == 2)
                    {
                        item.Date = DateTime.ParseExact(titleTemp[0].Replace(" ", ""), "dd/MM/yyyy", CultureInfo.InvariantCulture);
                        item.Title = WebUtility.HtmlDecode(titleTemp[1]);
                        item.Content = htmlTemp.DocumentNode.SelectNodes("//div[@class='tbBoxContent']")[0].InnerHtml;
                        item.ContentString = htmlTemp.DocumentNode.SelectNodes("//div[@class='tbBoxContent']")[0].InnerText;
                    }
                    else
                    {
                        item.Title = WebUtility.HtmlDecode(title);
                        item.Content = htmlTemp.DocumentNode.SelectNodes("//div[@class='tbBoxContent']")[0].InnerHtml;
                        item.ContentString = htmlTemp.DocumentNode.SelectNodes("//div[@class='tbBoxContent']")[0].InnerText;
                    }

                    result.Add(item);
                }
            }
            catch
            {
                result.Clear();
                result = null;
            }

            return result;
        }

        public static List<NewsGlobalItem> GetNewsGlobal(int page = 1, string query = null)
        {
            return GetNews(NewsType.Global, page, query);
        }

        public static List<NewsGlobalItem> GetNewsSubject(int page = 1, string query = null)
        {
            return GetNews(NewsType.Subject, page, query);
        }
    }
}
